### PR TITLE
init MVP

### PR DESCRIPTION
**Pull request by hungld-1302**
## Related Tickets
- [#Task 54284: Apply MVP Structure](https://edu-redmine.sun-asterisk.vn/issues/54284)
## WHAT
- Apply MVP Structure
## Review Checklist

Category | View Point | Description | Self review | Reviewer2 (name)
--- | --- | --- | --- | ---
Conventions | Sun* coding conventions | https://sal.vn/WtYPAy |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Conventions | Kotlin coding conventions | https://kotlinlang.org/docs/coding-conventions.html |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Redmine | Sun* Redmine working process  | https://sal.vn/piPoQn |<li>- [ ] yes</li>|<li>- [ ] yes</li>  

## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/java/com/sildev/tvshows/data/repository/TVShowRepository.kt [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package com.sildev.tvshows.data.repository
+
+import com.sildev.tvshows.data.model.TVShow
+import com.sildev.tvshows.data.repository.source.TVShowDataSource
+import com.sildev.tvshows.data.repository.source.remote.OnResultListener
+
+class TVShowRepository private constructor(
```
Comment (ndt0409) [inline app/src/main/java/com/sildev/tvshows/data/repository/TVShowRepository.kt]: không cần private constructor nhé

File: app/src/main/java/com/sildev/tvshows/data/model/TVShowDetail.kt [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+data class TVShowDetail(
+    val url: String = "",
+    val description: String = "",
+    val runtime: Int = 0,
+    val imagePath: String = "",
+    val rating: String = "",
+    val genres: MutableList<String> = mutableListOf(),
+    val pictures: MutableList<String> = mutableListOf(),
+    val episodes: JSONArray = JSONArray("")
```
Comment (minhnn-1279) [inline app/src/main/java/com/sildev/tvshows/data/model/TVShowDetail.kt]: chỗ này sao a ko lưu list object mà lưu JsonArray làm j nhỉ :v

File: app/src/main/java/com/sildev/tvshows/data/repository/source/remote/fetchjson/GetJsonFromUrl.kt [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+    private fun callAPI() {
+        mExecutor.execute {
+            val responseJson =
+                getJson(urlString)
+            data = ParseDataWithJson().parseJsonToData(JSONObject(responseJson), keyEntity) as? T
+            mHandler.post {
+                try {
+                    data?.let { listener.onSuccess(it) }
+                } catch (e: java.lang.NullPointerException) {
```
Comment (minhnn-1279) [inline app/src/main/java/com/sildev/tvshows/data/repository/source/remote/fetchjson/GetJsonFromUrl.kt]:
```suggestion
                } catch (exception: Exception) {
```

File: app/src/main/java/com/sildev/tvshows/data/repository/source/remote/OnResultListener.kt [hunk 4]
```diff
@@ -0,0 +1,6 @@
+package com.sildev.tvshows.data.repository.source.remote
+
+interface OnResultListener<T> {
+    fun onSuccess(data: T)
+    fun onError(exception: java.lang.Exception?)
```
Comment (minhnn-1279) [inline app/src/main/java/com/sildev/tvshows/data/repository/source/remote/OnResultListener.kt]:
```suggestion
    fun onError(exception: Exception?)
```

File: app/src/main/java/com/sildev/tvshows/data/repository/source/remote/fetchjson/GetJsonFromUrl.kt [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+import com.sildev.tvshows.data.repository.source.remote.OnResultListener
+import org.json.JSONObject
+import java.io.BufferedReader
+import java.io.InputStreamReader
+import java.net.HttpURLConnection
+import java.net.URL
+import java.util.concurrent.Executor
+import java.util.concurrent.Executors
+
+class GetJsonFromUrl<T> constructor(
```
Comment (minhnn-1279) [inline app/src/main/java/com/sildev/tvshows/data/repository/source/remote/fetchjson/GetJsonFromUrl.kt]:
```suggestion
class GetJsonFromUrl<T> (
```

File: app/src/main/java/com/sildev/tvshows/screen/listtvshow/MainPresenter.kt [hunk 6]
```diff
@@ -0,0 +1,35 @@
+package com.sildev.tvshows.screen.listtvshow
+
+import com.sildev.tvshows.data.model.TVShow
+import com.sildev.tvshows.data.repository.TVShowRepository
+import com.sildev.tvshows.data.repository.source.remote.OnResultListener
+
+class MainPresenter internal constructor(private val tvShowRepository: TVShowRepository?) :
```
Comment (minhnn-1279) [inline app/src/main/java/com/sildev/tvshows/screen/listtvshow/MainPresenter.kt]:
```suggestion
class MainPresenter (private val tvShowRepository: TVShowRepository?) :
```

File: app/src/main/AndroidManifest.xml [hunk 7]
```diff
@@ -2,6 +2,7 @@
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
     xmlns:tools="http://schemas.android.com/tools">
 
+    <uses-permission android:name="android.permission.INTERNET" />
     <application
```
Comment (vietnh-1109) [inline app/src/main/AndroidManifest.xml]:
```suggestion
    <uses-permission android:name="android.permission.INTERNET" />
    
    <application
```

File: app/src/main/java/com/sildev/tvshows/data/repository/TVShowRepository.kt [hunk 8]
```diff
@@ -0,0 +1,25 @@
+package com.sildev.tvshows.data.repository
+
+import com.sildev.tvshows.data.model.TVShow
+import com.sildev.tvshows.data.repository.source.TVShowDataSource
+import com.sildev.tvshows.data.repository.source.remote.OnResultListener
+
+class TVShowRepository constructor(
+    private val remote: TVShowDataSource.Remote, private val local: TVShowDataSource.Local
+) : TVShowDataSource.Remote, TVShowDataSource.Local {
+    override fun getTVShows(listener: OnResultListener<MutableList<TVShow>>) {
```
Comment (vietnh-1109) [inline app/src/main/java/com/sildev/tvshows/data/repository/TVShowRepository.kt]: chú ý lần cách 1 dòng ra nhé 
```suggestion
) : TVShowDataSource.Remote, TVShowDataSource.Local {

    override fun getTVShows(listener: OnResultListener<MutableList<TVShow>>) {
```

File: app/src/main/java/com/sildev/tvshows/data/repository/source/local/TVShowLocalDataSource.kt [hunk 9]
```diff
@@ -0,0 +1,18 @@
+package com.sildev.tvshows.data.repository.source.local
+
+import com.sildev.tvshows.data.model.TVShow
+import com.sildev.tvshows.data.repository.source.TVShowDataSource
+import com.sildev.tvshows.data.repository.source.remote.OnResultListener
+
+class TVShowLocalDataSource : TVShowDataSource.Local {
+    override fun getFavouriteTVShows(listener: OnResultListener<MutableList<TVShow>>) {
```
Comment (vietnh-1109) [inline app/src/main/java/com/sildev/tvshows/data/repository/source/local/TVShowLocalDataSource.kt]:
```suggestion
class TVShowLocalDataSource : TVShowDataSource.Local {

    override fun getFavouriteTVShows(listener: OnResultListener<MutableList<TVShow>>) {
```

File: app/src/main/java/com/sildev/tvshows/data/repository/source/remote/TVShowRemoteDataSource.kt [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+import com.sildev.tvshows.data.model.TVShow
+import com.sildev.tvshows.data.model.TVShowEntry
+import com.sildev.tvshows.data.repository.source.TVShowDataSource
+import com.sildev.tvshows.data.repository.source.remote.fetchjson.GetJsonFromUrl
+import com.sildev.tvshows.utils.BASE_POPULAR_TV_SHOW
+import com.sildev.tvshows.utils.BASE_URL
+
+class TVShowRemoteDataSource : TVShowDataSource.Remote {
+    override fun getTVShows(listener: OnResultListener<MutableList<TVShow>>) {
```
Comment (vietnh-1109) [inline app/src/main/java/com/sildev/tvshows/data/repository/source/remote/TVShowRemoteDataSource.kt]:
```suggestion
class TVShowRemoteDataSource : TVShowDataSource.Remote {

    override fun getTVShows(listener: OnResultListener<MutableList<TVShow>>) {
```

File: app/src/main/java/com/sildev/tvshows/data/repository/source/remote/fetchjson/ParseDataWithJson.kt [hunk 11]
```diff
@@ -0,0 +1,29 @@
+package com.sildev.tvshows.data.repository.source.remote.fetchjson
+
+import com.sildev.tvshows.data.model.TVShowEntry
+import org.json.JSONObject
+
```
Comment (vietnh-1109) [inline app/src/main/java/com/sildev/tvshows/data/repository/source/remote/fetchjson/ParseDataWithJson.kt:5]: remove

File: app/src/main/res/values/colors.xml [hunk 12]
```diff
@@ -7,8 +7,6 @@
     <color name="teal_700">#FF018786</color>
     <color name="black">#FF000000</color>
     <color name="white">#FFFFFFFF</color>
```
Comment (vietnh-1109) [inline app/src/main/res/values/colors.xml]: thêm tiền tố color_ vào đầu nhé

**Pull request merged**